[inductor] Add ir.WelfordReduction with multiple outputs

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104725

This replaces `var_unnormalized` reduction type with `welford_reduce` which takes the input data and outputs not just the variance, but also the mean and weights which account for the full welford accumulator state. Thus we can avoid re-computing the mean, and we now have enough information to create a multilayer reduction which I implement here by adding a second reduction type called `welford_combine` which reduces over all three inputs simultaneously.

Multi-layer support is particularly important as normalization operators like BatchNorm are being split in many timm models, which meant `var_unnormalized` had to fall back to two-pass variance calculation. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

peterbell10 committed:
WIP: [inductor] Add ir.WelfordReduction with multiple outputs

[ghstack-poisoned]

lezcano commented on torch/lib/libshm/CMakeLists.txt:
```diff
@@ -29,7 +29,12 @@ set_target_properties(shm PROPERTIES
   IMPORT_PREFIX "lib"
   CXX_STANDARD 17)
-target_link_libraries(shm PUBLIC torch)
+target_link_libraries(shm PUBLIC
+  torch
+  CUDA::cupti
+  CUDA::cusparse
+  ${CUDNN_LIBRARY_PATH}
+)
```
ooops

lezcano replied:
https://www.youtube.com/watch?v=CduA0TULnow

peterbell10 committed:
Update on "WIP: [inductor] Add ir.WelfordReduction with multiple outputs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: [inductor] Add ir.WelfordReduction with multiple outputs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: [inductor] Add ir.WelfordReduction with multiple outputs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: [inductor] Add ir.WelfordReduction with multiple outputs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: [inductor] Add ir.WelfordReduction with multiple outputs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: [inductor] Add ir.WelfordReduction with multiple outputs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

peterbell10 committed:
Rebase on "WIP: [inductor] Add ir.WelfordReduction with multiple outputs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: [inductor] Add ir.WelfordReduction with multiple outputs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: [inductor] Add ir.WelfordReduction with multiple outputs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

peterbell10 commented:
This gives a nice 10% speedup in many timm models:
https://hud.pytorch.org/benchmark/timm_models/inductor_no_cudagraphs?startTime=Wed%2C%2026%20Jul%202023%2014%3A01%3A20%20GMT&stopTime=Wed%2C%2002%20Aug%202023%2014%3A01%3A20%20GMT&granularity=hour&mode=training&dtype=amp&lBranch=gh/peterbell10/582/orig&lCommit=1ee353815cda0a504348075fc81c8f3586ab75ff&rBranch=main&rCommit=0af3203c728278920f6e64b22f63181c6f88628a

lezcano commented on torch/_inductor/ir.py:
```diff
@@ -1079,6 +1086,289 @@ def intermediate_fn(index, reduction_index):
+    return 3 if "welford" in reduction_type else 1
+
+
+class WelfordReduction(Reduction):
```
Wouldn't it be possible to generalise `Reduction` to be able to handle multiple inputs / outputs? It's a bit concerning how much repeated code we have here...

peterbell10 committed:
Rebase and small multilayer refactor on "[inductor] Add ir.WelfordReduction with multiple outputs"


This replaces `var_unnormalized` reduction type with `welford_reduce` which takes the input data and outputs not just the variance, but also the mean and weights which account for the full welford accumulator state. Thus we can avoid re-computing the mean, and we now have enough information to create a multilayer reduction which I implement here by adding a second reduction type called `welford_combine` which reduces over all three inputs simultaneously.

Multi-layer support is particularly important as normalization operators like BatchNorm are being split in many timm models, which meant `var_unnormalized` had to fall back to two-pass variance calculation. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

lezcano reviewed (approved):
I'm not the biggest fan of all the duplicate code in ir.py. It'd be nice to have a go at some point at having a Reduction IR node that handles single and multiple outputs.

At any rate, great PR twisting inductor's arm to make it support this reduction via the `store_reduction` split.

peterbell10 committed:
Respond to comments and remove var_unnormalized references on "[inductor] Add ir.WelfordReduction with multiple outputs"


This replaces `var_unnormalized` reduction type with `welford_reduce` which takes the input data and outputs not just the variance, but also the mean and weights which account for the full welford accumulator state. Thus we can avoid re-computing the mean, and we now have enough information to create a multilayer reduction which I implement here by adding a second reduction type called `welford_combine` which reduces over all three inputs simultaneously.

Multi-layer support is particularly important as normalization operators like BatchNorm are being split in many timm models, which meant `var_unnormalized` had to fall back to two-pass variance calculation. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/index_propagation.py:
```diff
@@ -171,15 +174,22 @@ def unwrap(self, a: Union[Any, IndexPropVar]) -> Any:
     def propagate_sympy(
         self, name: str, args: Tuple, kwargs: Dict[str, Any]
-    ) -> Union[TypedExpr, IndexPropVar]:
+    ) -> IndexPropVar:
```
This function never returns a tuple at the moment, so the typing here is correct.

peterbell10 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1407,6 +1412,16 @@ def reduction(self, dtype, src_dtype, reduction_type, value):
         reduction_sizes[-1] = ":"
 
+        # The shape of the reduction dimension changes the output, so
+        # explicitly broadcast the input to the expected shape.
```
Added an example that should be clearer. Before this change
```python
tmp0 = ops.constant(1, torch.int64)
tmp1 = ops.reduction(torch.int64, torch.int64, "sum", tmp0)
```
becomes
```python
tmp0 = tl.full([1, 1], 1, tl.int64)
tmp1 = tl.sum(tmp0, 1)[:, None]
```
which is wrong because summing over 1 element is not equivalent to summing over `RBLOCK` elements.

peterbell10 commented:
@pytorchbot merge

peterbell10 committed:
Update on "[inductor] Add ir.WelfordReduction with multiple outputs"


This replaces `var_unnormalized` reduction type with `welford_reduce` which takes the input data and outputs not just the variance, but also the mean and weights which account for the full welford accumulator state. Thus we can avoid re-computing the mean, and we now have enough information to create a multilayer reduction which I implement here by adding a second reduction type called `welford_combine` which reduces over all three inputs simultaneously.

Multi-layer support is particularly important as normalization operators like BatchNorm are being split in many timm models, which meant `var_unnormalized` had to fall back to two-pass variance calculation. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

peterbell10 committed:
rebase on "[inductor] Add ir.WelfordReduction with multiple outputs"


This replaces `var_unnormalized` reduction type with `welford_reduce` which takes the input data and outputs not just the variance, but also the mean and weights which account for the full welford accumulator state. Thus we can avoid re-computing the mean, and we now have enough information to create a multilayer reduction which I implement here by adding a second reduction type called `welford_combine` which reduces over all three inputs simultaneously.

Multi-layer support is particularly important as normalization operators like BatchNorm are being split in many timm models, which meant `var_unnormalized` had to fall back to two-pass variance calculation. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
